subtests/docker_cli/liverestore/liverestore.py

```diff
@@ -0,0 +1,200 @@
+        """
+        Start a container. We only need it briefly, until we (the test,
+        running in host space) can cd into its root filesystem. Container
+        will spin idly until a semaphore file is removed.
+        """
+        c_name = self.stuff['dc'].get_unique_name()
+
+        fin = DockerImage.full_name_from_defaults(self.config)
+        subargs = ['--detach', '--name=' + c_name, fin,
+                   'bash -c "echo READY;'
+                   ' while :; do date +%s; sleep 1;done"']
+        dkrcmd = AsyncDockerCmd(self, 'run', subargs)
+        dkrcmd.execute()
+        dkrcmd.wait_for_ready(c_name)
+        self.stuff['container_name'] = c_name
+
+    def _verify_that_date_changes(self):
+        """
+        Verify that container process is running, by running 'docker log'
+        on it sequentially and making sure that (1) we get new output,
+        and (2) the new output corresponds to current time.
+        """
+        log_0 = self._container_log()
+
+        # Wait up to 5 seconds for output to change
+        end_at = time.time() + 5
+        while time.time() <= end_at:
+            log_1 = self._container_log()
+            if log_1 != log_0:
+                break
+            time.sleep(1)
+
+        self.failif(len(log_0) == len(log_1),
+                    "Container log line count did not grow")
+
+    def _container_log(self):
+        """
+        Returns container log as a list of strings
+        """
+        c_name = self.stuff['container_name']
+        result = mustpass(DockerCmd(self, "logs", [c_name]).execute())
+        return result.stdout.strip().split("\n")
+
+    def _container_pid(self):
+        dc = self.stuff['dc']
+        inspect = dc.json_by_name(self.stuff['container_name'])
+        return int(inspect[0]['State']['Pid'])
+
+    def _container_is_child_of_docker_daemon(self):
+        """
+        Returns True if our container's PID is a child of the docker daemon.
+        """
+        dockerd_pid = docker_daemon.pid()
+
+        pid = self._container_pid()
+        while pid != 1:
+            pid = self._ppid(pid)
+            if pid == dockerd_pid:
+                return True
+        return False
+
+    @staticmethod
+    def _ppid(pid):
+        """ returns parent PID of given PID """
+        return int(utils.run('ps -o ppid= -p %d' % pid).stdout.strip())
+
+    def postprocess(self):
+        super(liverestore, self).postprocess()
+
+        # Run 'docker ps', confirm that container is listed and is active
+        dc = self.stuff['dc']
+        docker_ps = dc.list_containers_with_name(self.stuff['container_name'])
+        self.failif(not docker_ps,
+                    "Container not in 'docker ps' output after restart")
+        self.failif_not_in("Up ", docker_ps[0].status,
+                           "container status as reported by 'docker ps'")
+
+        # Confirm that container is still running (spitting out dates)
+        self._verify_that_date_changes()
```

Comment:
Probably why the name is in parenthesis, though you could count the fields from the end as well to avoid extra parsing.  An implied use-case of live-restore is also that the container continues to do work while disconnected from the docker daemon.  IMHO, checking the CPU counters for the process is a simpler proof than date-based check.  OTOH, it's such a fine-line, it's not worth making a fuss over, what you have is fine too.